sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala

```diff
@@ -439,6 +439,13 @@ case class MapObjects(
       s"boolean ${loopVar.isNull} = ${genInputData.isNull} || ${loopVar.value} == null;"
     }
 
+    // If lambdaFunction is WrapOption, we will not determine null or not based on the
+    // value of loopVar.isNull, because WrapOption will return None for null.
+    val isWrapOption = lambdaFunction match {
```

Comment:
@marmbrus I have refactored this part and removed the check of WrapOption now. Please take a look it if it is better now. Thanks.